File: src/main/java/io/georocket/util/io/BufferWriteStream.java
package io.georocket.util.io;

import io.vertx.core.AsyncResult;
import io.vertx.core.Future;
import io.vertx.core.Handler;
import io.vertx.core.Promise;
import io.vertx.core.buffer.Buffer;
import io.vertx.core.streams.WriteStream;

/**
 * A {@link WriteStream} that collects all written data in a {@link Buffer}
 * @author Michel Kraemer
 */
public class BufferWriteStream implements WriteStream<Buffer> {
  private final Buffer buf = Buffer.buffer();
  
  @Override
  public WriteStream<Buffer> exceptionHandler(Handler<Throwable> handler) {
    // exceptions cannot happen
    return this;
  }

  @Override
  public Future<Void> write(Buffer data) {
    Promise<Void> promise = Promise.promise();
    write(data, promise);
    return promise.future();
  }

  @Override
  public void write(Buffer data, Handler<AsyncResult<Void>> handler) {
    buf.appendBuffer(data);
    if (handler != null) {
      handler.handle(Future.succeededFuture());
    }
  }

  @Override
  public WriteStream<Buffer> setWriteQueueMaxSize(int maxSize) {
    return this; // ignore
  }

  @Override
  public boolean writeQueueFull() {
    return false; // never full
  }

  @Override
  public WriteStream<Buffer> drainHandler(Handler<Void> handler) {
    // we don't need a drain handler because we're never full
    return this;
  }
  
  @Override
  public Future<Void> end() {
    Promise<Void> promise = Promise.promise();
    end(promise);
    return promise.future();
  }

  @Override
  public void end(Handler<AsyncResult<Void>> handler) {
    if (handler != null) {
      handler.handle(Future.succeededFuture());
    }
  }
  
  /**
   * @return the buffer
   */
  public Buffer getBuffer() {
    return buf;
  }
}

File: src/main/kotlin/io/georocket/util/io/BufferWriteStream.kt
package io.georocket.util.io

import io.vertx.core.AsyncResult
import io.vertx.core.Future
import io.vertx.core.Handler
import io.vertx.core.Promise
import io.vertx.core.buffer.Buffer
import io.vertx.core.streams.WriteStream

/**
 * A [WriteStream] that collects all written data in a [Buffer]
 * @author Michel Kraemer
 */
class BufferWriteStream : WriteStream<Buffer> {
  /**
   * @return the buffer
   */
  val buffer = Buffer.buffer()

  override fun exceptionHandler(handler: Handler<Throwable>): WriteStream<Buffer> {
    // exceptions cannot happen
    return this
  }

  override fun write(data: Buffer): Future<Void> {
    val promise = Promise.promise<Void>()
    write(data, promise)
    return promise.future()
  }

  override fun write(data: Buffer, handler: Handler<AsyncResult<Void>>) {
    buffer.appendBuffer(data)
    handler.handle(Future.succeededFuture())
  }

  override fun setWriteQueueMaxSize(maxSize: Int): WriteStream<Buffer> {
    return this // ignore
  }

  override fun writeQueueFull(): Boolean {
    return false // never full
  }

  override fun drainHandler(handler: Handler<Void>): WriteStream<Buffer> {
    // we don't need a drain handler because we're never full
    return this
  }

  override fun end(): Future<Void> {
    val promise = Promise.promise<Void>()
    end(promise)
    return promise.future()
  }

  override fun end(handler: Handler<AsyncResult<Void>>) {
    handler.handle(Future.succeededFuture())
  }
}
